### PR TITLE
FIR checker/IDE: Add checker and quickfix for VAL_WITH_SETTER

### PR DESCRIPTION
**Pull request by punzki**
The changes to FIR building were necessary as no `FirPropertyAccessor` was created for a setter inside a `val` property.

### CODE REVIEW

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt [hunk 1]
```diff
@@ -38,16 +38,14 @@ internal fun checkExpectDeclarationVisibilityAndBody(
     }
 }
 
-internal fun checkPropertyInitializer(
+internal fun checkProperty(
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:41]: Perhaps this is quite personal/cosmetic issue, but I still prefer to separate stuff in some logical chunk, like checkPropertyInitializer, checkPropertyAccessors, etc., then you can still use this, checkProperty, as a one entry point.

File: compiler/fir/raw-fir/light-tree2fir/src/org/jetbrains/kotlin/fir/lightTree/converter/DeclarationsConverter.kt [hunk 2]
```diff
@@ -1017,16 +1017,15 @@ class DeclarationsConverter(
...
-                    this.setter =
-                        if (isVar) {
-                            convertedAccessors.find { it.isSetter }
-                                ?: FirDefaultPropertySetter(
-                                    null, session, FirDeclarationOrigin.Source, returnType, propertyVisibility
-                                ).also {
-                                    currentDispatchReceiverType()?.lookupTag?.let { lookupTag ->
-                                        it.containingClassAttr = lookupTag
-                                    }
+                    this.setter = convertedAccessors.find { it.isSetter }
```
Comment (jsjeon) [inline compiler/fir/raw-fir/light-tree2fir/src/org/jetbrains/kotlin/fir/lightTree/converter/DeclarationsConverter.kt:1020]: Minor, but isn't it still worth leaving some comments here (and corresponding part in `RawFirBuilder`) that we still need to build a setter for `val` to report `VAL_WITH_SETTER`?

File: compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt [hunk 3]
```diff
@@ -345,68 +345,73 @@ class RawFirBuilder(
...
-                val bodyWithContractDescription = this@toFirPropertyAccessor.buildFirBody()
-                this.body = bodyWithContractDescription.first
-                val contractDescription = outerContractDescription ?: bodyWithContractDescription.second
-                contractDescription?.let {
-                    this.contractDescription = it
-                }
-            }.also {
-                currentDispatchReceiverType()?.lookupTag?.let { lookupTag ->
-                    it.containingClassAttr = lookupTag
+                isGetter || property.isVar -> {
```
Comment (tgeng) [inline compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:391]: IIUC there is no logic change in this file. Is this change just to improve the readability?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt [hunk 4]
```diff
@@ -39,6 +39,10 @@ object FirMemberPropertyChecker : FirRegularClassChecker() {
         // If multiple (potentially conflicting) modality modifiers are specified, not all modifiers are recorded at `status`.
         // So, our source of truth should be the full modifier list retrieved from the source.
         val modifierList = with(FirModifierList) { property.source.getModifierList() }
```
[the remark this answers is not in the document]
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt:41]: It's kind of legacy :\ All those modifier retrieval utils _were_ `private` utils in modifier checkers, so we need `with` to access those (literally) with the receiver of the containing class type.

Well, at least we can avoid the following long import:
`import org.jetbrains.kotlin.fir.analysis.checkers.declaration.FirModifierList.Companion.getModifierList`
:p but yeah, either should be fine.

File: compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt (hunk 3 above)
Comment (jsjeon) [reply compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:391]: This is really subtle, and thus need some code comments (as I described below).

Previously, if the receiver of `KtPropertyAccessor` type is not given or it doesn't have a body, we created a default accessor. For the remaining case---`KtPropertyAccessor` is given _and_ the body is also declared, we created a property accessor.

Now, with this change, we did the latter first, which is a clear case I think. Then, this check (`isGetter || property.isVar`) prevent us from creating a default accessor if it is setter inside `val` property. But, you said,
> The changes to FIR building were necessary as no `FirPropertyAccessor` was created for a setter inside a `val` property.

Thus, this change looks opposite; or the condition is somewhat incorrect; or maybe as @tgeng mentioned, you just tried to clean up? I think line 1277 below is the only necessary change?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt [hunk 5]
```diff
@@ -39,6 +39,10 @@ object FirMemberPropertyChecker : FirRegularClassChecker() {
         // If multiple (potentially conflicting) modality modifiers are specified, not all modifiers are recorded at `status`.
```
[the remark this answers is not in the document]
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt:39]: Not sure what you're asking me to do here. Are you asking me to fix this bug here too?

File: compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt (hunk 3 above)
Comment (punzki) [reply compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:391]: @jsjeon is correct, the only thing I added is to add `isGetter || property.isVar` when creating a default accessor, else return `null` (no accessor for setter in `val` property). And the rest is cosmetic, rearranging so the conditions are easier to read and smartcasting to not-null in the first branch works.

I will add some comments as suggested.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt (hunk 1 above)
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:41]: I considered that but there's only one extra check on the accessor, and the current function also does some examination of the accessors too so it's not clear what each function would be responsible for.

Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:41]: > the current function also does some examination of the accessors too

If you refer to `hasAccessorImplementation`, it's used to differentiate diagnostics for uninitialized properties. So, in terms of reported diagnostics, all were related to property initializer.

But, again, this is cosmetic stuff. When we have more diagnostics for accessors, we can revisit those and make them as a separate util like I suggested.

File: compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt (hunk 3 above)
Comment (jsjeon) [reply compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:393]: With comments added, now things look much clearer! 👍

**Review by jsjeon — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt:39]: Nope, I don't expect it's fixed in this PR. Just want to bring it to attention. Maybe add a TODO?

[the remark this answers is not in the document]
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt:39]: Actually, it looks like this should be covered by `ABSTRACT_PROPERTY_IN_PRIMARY_CONSTRUCTOR_PARAMETERS`. So maybe the current code is intentional before this diagnostic code is implemented?

File: compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt (hunk 3 above)
Comment (tgeng) [reply compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:393]: Just to make sure I understand it correctly. So the changes in this file are indeed cosmetic and introduce no behavior change, right?

Comment (punzki) [reply compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:393]: Sorry it is not clear. There _is_ a behavior change from the added condition and the `else null` which is new (previously the function returns not-nullable), as explained in my previous post.

Comment (tgeng) [reply compiler/fir/raw-fir/psi2fir/src/org/jetbrains/kotlin/fir/builder/RawFirBuilder.kt:393]: Yeah, but the caller is changed too. And it seems to me that all the changes in this file combined cause no behavior changes.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt (hunk 1 above)
Comment (darthorimar) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:44]: I'm for splitting that function too. We can at least extract the new check to `fun checkForValpropertyWithSetter(...)`

**Review by darthorimar — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt (hunk 1 above)
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:44]: I split up the function into `checkPropertyInitializer` (original) and `checkPropertyAccessors` (so we don't have to rename it again in the future if the name was too specific)

**Review by mglukhikh — APPROVED** [review]

**Comment by darthorimar** [comment]
Thank you!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt (hunk 1 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:44]: I'm even removing this at https://github.com/JetBrains/kotlin/pull/4178 as per https://github.com/JetBrains/kotlin/pull/4142#discussion_r587280893.

**Pull request merged**